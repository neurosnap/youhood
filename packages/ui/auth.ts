import * as h from 'react-hyperscript';
import { Component } from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';

import { actions, selectors as authSelectors } from '@youhood/auth';
const { signIn, register } = actions;
const { getAuthError } = authSelectors;
import { selectors } from '@youhood/user';
const { getCurrentUserId } = selectors;
import { AuthPayload, AuthError } from '@youhood/auth/types';
import { UserId } from '@youhood/user/types';

import { State } from '@youhood/web-app/types';
import {
  DropdownMenuButton,
  ErrorText,
  Input,
  Buttons,
  SignInMenuEl,
  DropdownMenuContainer,
  SigninMsgBase,
} from './ui';

type OnClick = (props: AuthPayload) => void;
interface IAuth {
  error?: AuthError;
  children?: any;
  onClick: OnClick;
  currentUserId: UserId;
  buttonText: string;
}

interface IAuthState {
  email: string;
  password: string;
}

export class Auth extends Component<IAuth, IAuthState> {
  static defaultProps = {
    buttonText: '',
    onClick: () => {},
    currentUserId: '',
    error: '',
  };

  state = {
    email: '',
    password: '',
  };

  handleEmail = (event: React.FormEvent<HTMLInputElement>) => {
    this.setState({ email: event.currentTarget.value });
  };

  handlePassword = (event: React.FormEvent<HTMLInputElement>) => {
    this.setState({ password: event.currentTarget.value });
  };

  handleClick = () => {
    const { email, password } = this.state;
    const { onClick, currentUserId } = this.props;
    onClick({ email, password, currentUserId });
  };

  render() {
    const { email, password } = this.state;
    const { error, buttonText } = this.props;
    const shouldDisableButton = !email || !password;

    return h(SignInMenuEl, [
      h(Input, {
        className: 'signin-email',
        type: 'text',
        placeholder: 'email address',
        value: email,
        onChange: this.handleEmail,
      }),
      h(Input, {
        className: 'signin-pass',
        type: 'password',
        placeholder: 'password',
        value: password,
        onChange: this.handlePassword,
      }),
      error ? h(ErrorText, error) : h(SigninMsgBase),
      h(Buttons, [
        h(
          DropdownMenuButton,
          {
            className: 'signin-btn',
            onClick: this.handleClick,
            disabled: shouldDisableButton,
          },
          buttonText,
        ),
      ]),
    ]);
  }
}

interface IAuthMenu {
  error: AuthError;
  onClick: OnClick;
  currentUserId: UserId;
}

export const createAuthMenu = (buttonText: string) => ({
  onClick,
  currentUserId,
  error,
}: IAuthMenu) => h(Auth, { onClick, currentUserId, error, buttonText });

export const SignInMenu = createAuthMenu('Sign In');
export const RegisterMenu = createAuthMenu('Register');

const mapStateToProps = (state: State) => ({
  currentUserId: getCurrentUserId(state),
  error: getAuthError(state),
});

const createConnectForAuth = (component: any, action: Function) =>
  connect(
    mapStateToProps,
    (dispatch: Function) => ({
      onClick: (payload: AuthPayload) => dispatch(action(payload)),
    }),
  )(component as any);

export const SignInMenuConn = createConnectForAuth(SignInMenu, signIn);
export const RegisterMenuConn = createConnectForAuth(RegisterMenu, register);

type Tab = 'Sign In' | 'Register';
interface IAuthMenuState {
  tabSelected: Tab;
}

const TabSelected = styled.div`
  text-decoration: underline;
`;
const TabInactive = styled.div`
  cursor: pointer;
  color: #ccc;

  :hover {
    color: #fff;
  }
`;
const TabMenu = styled.div`
  display: flex;
  justify-content: space-evenly;
  align-items: center;
  height: 36px;
`;

export default class AuthMenu extends Component<null, IAuthMenuState> {
  state = {
    tabSelected: 'Sign In' as Tab,
  };

  setTab = (tab: Tab) => {
    const { tabSelected } = this.state;
    if (tab === tabSelected) return;
    this.setState({ tabSelected: tab });
  };

  render() {
    const { tabSelected } = this.state;
    const tabNames = ['Sign In', 'Register'];
    const tabs = tabNames.map((tab: Tab) => {
      const onClick = () => this.setTab(tab);
      const TabView = tab === tabSelected ? TabSelected : TabInactive;
      return h(TabView, { onClick, key: tab }, tab);
    });
    const Menu = tabSelected === 'Sign In' ? SignInMenuConn : RegisterMenuConn;

    return h(DropdownMenuContainer, [h(TabMenu, tabs), h(Menu)]);
  }
}
