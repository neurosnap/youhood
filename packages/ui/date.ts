function addZero(num: number): string {
  if (num < 10) {
    return `0${num}`;
  }

  return `${num}`;
}

function militaryToTime(hour: number) {
  if (hour > 12) {
    return Math.abs(12 - hour);
  }

  return hour;
}

function getAmOrPm(hour: number) {
  return hour > 12 ? 'pm' : 'am';
}

function getLocalTime(date: Date): string {
  const rawHours = date.getHours();
  const hours = addZero(militaryToTime(rawHours));
  const amOrPm = getAmOrPm(rawHours);
  const time = `${hours}:${addZero(date.getMinutes())} ${amOrPm}`;
  return time;
}

export function formatDateTime(dateStr: string) {
  if (!dateStr) {
    return ['', ''];
  }

  const date = new Date(dateStr);

  const month = addZero(date.getMonth() + 1);
  const tDate = addZero(date.getDate());
  const d = `${month}/${tDate}/${date.getFullYear()}`;

  const time = getLocalTime(date);
  return [d, time];
}

export function formatDate(date: string) {
  const [d, time] = formatDateTime(date);
  return `${d} ${time}`;
}
