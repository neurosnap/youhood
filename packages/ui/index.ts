import Overlay from './overlay';
import Navbar from './navbar';
import HoodBar from './hoodbar';

const smartComponents = {
  Overlay,
  Navbar,
  HoodBar,
};

export { smartComponents };
