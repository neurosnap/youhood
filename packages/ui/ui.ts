import styled from 'styled-components';

const successColor = '#28a745';
const errorColor = '#dc3545';
const primaryFontColor = '#fff';

// core

export const Header = styled.h1`
  font-size: 1.3rem;
`;

export const HeaderSmall = styled.h2`
  font-size: 1.2rem;
  margin-bottom: 0.5rem;
`;

export const Link = styled.a`
  margin: 0 0.2rem;
  color: ${primaryFontColor};
`;

export const LinkDanger = Link.extend`
  color: ${errorColor};
`;

export const LinkSuccess = Link.extend`
  color: ${successColor};
`;

export const InputBase = styled.input`
  margin: 0;
  outline: none;
`;

export const TextSmall = styled.div`
  font-size: 0.8rem;
  display: flex;
  justify-content: space-between;
`;

// derived

export const InputOverlay = InputBase.extend`
  padding: 0.375rem 0.75rem;
  color: #495057;
  background-color: #fff;
  transition: border-color 0.15s ease-in-out, box-shadow 0.15s ease-in-out;
  width: 100%;
`;

const OverlaySection = styled.div`
  background-color: #fff;
  border-radius: 0.3rem;
  padding: 0.6rem;
`;

export const HoodContainer = OverlaySection.extend`
  display: flex;
`;
export const HoodEditorContainer = OverlaySection;

export const HoodSelectionContainer = OverlaySection.extend`
  display: flex;
  flex-direction: column;
  margin-bottom: 0.5rem;
`;

export const HoodSelectionItem = styled.div`
  display: flex;
  align-items: center;
`;

export const HoodVisibility = styled.i`
  cursor: pointer;
`;

export const Votes = styled.div`
  padding: 0 0.5rem;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;

export const VoteUp = styled.i`
  cursor: pointer;
`;

export const Voted = VoteUp.extend`
  color: orange;
`;

export const Actions = styled.div`
  display: flex;
  justify-content: space-between;
`;

export const EditorActions = Actions.extend`
  margin-top: 0.5rem;
`;

export const HoodBaseItem = styled.div`
  cursor: pointer;
  padding: 0.3rem 0.4rem;
  border-radius: 0.3rem;
  flex: 1;
  margin-left: 0.3rem;
`;

export const HoodListItem = HoodBaseItem.extend`
  :hover {
    background-color: rgba(46, 139, 87, 0.5);
  }
`;

export const HoodListItemSelected = HoodBaseItem.extend`
  background-color: rgb(255, 250, 205);
`;

export const OverlayHeader = styled.div`
  font-size: 1rem;
  color: ${primaryFontColor};
  display: flex;
  align-items: center;
  margin-bottom: 0.6rem;
`;

export const OverlayContainer = styled.div`
  width: 90%;
  margin-top: 0.5rem;
`;

export const NavHover = styled.button`
  cursor: pointer;
  height: 100%;
  display: flex;
  align-items: center;
  padding: 0 0.4rem;
  margin: 0 0.4rem;
  background: inherit;
  color: inherit;
  border: none;
  font: inherit;
  cursor: pointer;
  outline: inherit;

  :hover:enabled {
    background-color: rgba(255, 255, 255, 0.25);
  }

  :disabled {
    color: #ccc;
  }
`;

export const HoodBarButton = NavHover.extend`
  font-size: 0.8rem;
`;

export const Input = InputBase.extend`
  margin: 0.4rem;
  padding: 0.4rem;
  border-radius: 0.1rem;
  height: 3rem;
  border: none;
  background-image: none;
  background-color: rgba(255, 255, 255, 0.15);
  box-shadow: none;
  width: 90%;
  color: ${primaryFontColor};
  font-size: 1rem;

  :hover {
    background-color: rgba(255, 255, 255, 0.25);
  }

  ::-webkit-input-placeholder {
    color: #eee;
  }
`;

export const SigninMsgBase = styled.div`
  height: 2rem;
`;

export const ErrorText = SigninMsgBase.extend`
  color: #721c24;
  background-color: #f8d7da;
  font-size: 0.8rem;
  padding: 0 0.5rem;
  width: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
`;

export const Buttons = styled.div`
  margin-top: 0.4rem;
  display: flex;
`;

export const Menu = NavHover.extend`
  height: 2rem;
`;

export const Nav = styled.div`
  display: flex;
  align-items: center;
  height: 56px;
  background: #4285f4;
  box-sizing: border-box;
  top: 0;
  width: 100vw;
  color: ${primaryFontColor};
`;

export const NavContent = styled.div`
  width: 100%;
  margin: 0 1.5rem;
  display: flex;
  align-items: center;
`;

export const NavLeft = styled.div`
  display: flex;
  align-items: center;
`;

export const NavRight = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  height: 2.2rem;
  width: 75%;
`;

export const Search = styled.div`
  width: 40%;
  height: 100%;
  background-color: rgba(255, 255, 255, 0.15);
  border-radius: 0.1rem;
  display: flex;
  align-items: center;

  :hover {
    background-color: rgba(255, 255, 255, 0.25);
  }
`;

export const SearchInput = InputBase.extend`
  font-size: 1rem;
  background: none;
  border: none;
  box-sizing: border-box;
  height: 100%;
  flex: 1;
  color: ${primaryFontColor};

  ::-webkit-input-placeholder {
    color: ${primaryFontColor};
  }
`;

export const SearchIcon = styled.i`
  margin: 0 0.5rem;
`;

export const Brand = styled.div`
  font-size: 1.1rem;
  margin: 0 1.5rem;
`;

export const SignInMenuEl = styled.div`
  height: 80%;
  width: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
`;

export const DropdownMenuContainer = styled.div`
  position: absolute;
  background: #4285f4;
  top: 46px;
  right: 9px;
  z-index: 401;
  width: 310px;
  height: 260px;
`;

export const DropdownMenuButton = NavHover.extend`
  height: 2rem;
`;

export const Points = styled.div`
  color: ${primaryFontColor};
  background-color: limegreen;
  border-radius: 0.3rem;
  padding: 0.1rem 0.3rem;
  font-size: 0.8rem;
`;

export const SignInContainer = styled.div`
  margin-left: 0;
  position: relative;
  height: 100%;
`;

export const SignInEl = styled.div`
  text-decoration: none;
  color: ${primaryFontColor};

  a {
    text-decoration: none;
    color: ${primaryFontColor};
  }
`;
