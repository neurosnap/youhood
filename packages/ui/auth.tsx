import * as React from 'react';
import styled from 'styled-components';

import theme from '@youhood/theme';

import { DropdownMenuContainer } from './ui';
import SignIn from './signin';
import Register from './register';

type Tab = 'Sign In' | 'Register';
interface IAuthMenuState {
  tabSelected: Tab;
}

const TabSelected = styled.div`
  color: ${theme.palette.font.primary};
  text-decoration: underline;
`;
const TabInactive = styled.div`
  cursor: pointer;
  color: ${theme.palette.font.disabled};

  :hover {
    color: ${theme.palette.font.primary};
  }
`;
const TabMenu = styled.div`
  display: flex;
  justify-content: space-evenly;
  align-items: center;
  height: 36px;
`;

export class AuthMenu extends React.Component<any, IAuthMenuState> {
  state = {
    tabSelected: 'Sign In' as Tab,
  };

  setTab = (tab: Tab) => {
    const { tabSelected } = this.state;
    if (tab === tabSelected) return;
    this.setState({ tabSelected: tab });
  };

  render() {
    const { tabSelected } = this.state;
    const tabNames: Tab[] = ['Sign In', 'Register'];
    const tabs = tabNames.map((tab: Tab) => {
      const onClick = () => this.setTab(tab);
      const TabView = tab === tabSelected ? TabSelected : TabInactive;
      return (
        <TabView onClick={onClick} key={tab}>
          {tab}
        </TabView>
      );
    });
    const Menu = tabSelected === 'Sign In' ? SignIn : Register;

    return (
      <div>
        <TabMenu>{tabs}</TabMenu>
        <Menu />
      </div>
    );
  }
}

export default () => (
  <DropdownMenuContainer>
    <AuthMenu />
  </DropdownMenuContainer>
);
