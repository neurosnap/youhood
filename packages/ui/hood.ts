import { Component } from 'react';
import { connect } from 'react-redux';
import * as h from 'react-hyperscript';

import { HoodProps, HoodId, SetHoodNamePayload, EditHoodPayload } from '@youhood/hood/types';
import { User, UserId } from '@youhood/user/types';
import {
  actionCreators,
  selectors,
} from '@youhood/hood';
const {
  deselectHood,
  setHoodName,
  editHood,
  saveHood,
} = actionCreators;
const { getHoodSelected } = selectors;
import { actionCreators as menuActionCreators } from '@youhood/menu';
const { hideMenu } = menuActionCreators;
import { selectors as userSelectors } from '@youhood/user';
const { getUserById, getCurrentUserId } = userSelectors;
import {
  selectors as voteSelectors,
  actionCreators as voteActionCreators,
} from '@youhood/vote';
const { getVoteCountByHood, didUserVoteOnHood } = voteSelectors;
const { vote, unvote } = voteActionCreators;
import { selectors as authSelectors } from '@youhood/auth';
const { isUserAuthenticated } = authSelectors;

import { State } from '@youhood/web-app/types';
import {
  HoodContainer,
  Votes,
  VoteUp,
  Voted,
  Actions,
  OverlayHeader,
  OverlayContainer,
} from './ui';

const noop = () => {};

interface Props {
  hood: HoodProps;
  hoodId: HoodId;
  user: User;
  show: boolean;
  canEdit: boolean;
  updateHoodName: Function;
  handleDeselectHood: Function;
  hideHoodOverlay: Function;
  edit: Function;
  save: Function;
  votes: number;
  userVoted: boolean;
  canUserVote: boolean;
  handleVote: Function;
  handleUnvote: Function;
  currentUserId: UserId;
}

interface DefaultProps {
  show: boolean;
  hood: HoodProps;
  user: User;
  canEdit: boolean;
  userVoted: boolean;
  canUserVote: boolean;
  votes: number;
  edit: Function;
  save: Function;
}

function getHoodId(hood: HoodProps): HoodId {
  if (!hood) return '';
  return hood.id;
}

function getHoodName(hood: HoodProps): string {
  if (!hood) return '';
  return hood.name;
}

export class HoodView extends Component {
  props: Props;

  static defaultProps: DefaultProps = {
    show: false,
    hood: null,
    user: {
      email: 'Unknown',
      id: '',
      createdAt: '',
      isTmp: true,
    },
    canEdit: false,
    userVoted: false,
    canUserVote: false,
    votes: 0,
    edit: noop,
    save: noop,
  };

  state = {
    name: '',
    editing: false,
  };

  componentWillMount() {
    const name = getHoodName(this.props.hood);
    this.setState({ name });
  }

  componentWillReceiveProps(nextProps: Props) {
    if (getHoodId(nextProps.hood) !== getHoodId(this.props.hood)) {
      this.setState({ name: getHoodName(nextProps.hood) });
    }
  }

  handleSave = () => {
    const { hood, updateHoodName, edit, save } = this.props;
    const { name } = this.state;
    const hoodId = hood.id;
    updateHoodName({ hoodId, name });
    this.setState({ editing: false });
    edit({ hoodId, edit: false });
    save(hoodId);
  }

  handleCancel = () => {
    const { edit, hood } = this.props;
    const { name, id } = hood;
    this.setState({
      editing: false,
      name,
    });
    edit({ hoodId: id, edit: false });
  }

  handleInput = (event: Event) => {
    const name = (<HTMLInputElement>event.target).value;
    this.setState({ name });
  }

  handleEdit = () => {
    const { edit, hood } = this.props;
    this.setState({ editing: true });
    edit({ hoodId: hood.id, edit: true });
  }

  render() {
    const {
      hoodId,
      show,
      user,
      canEdit,
      votes,
      userVoted,
      canUserVote,
      handleVote,
      handleUnvote,
      currentUserId,
    } = this.props;
    if (!show || !hoodId) return null;
    const { name, editing } = this.state;

    let actions = null;
    if (editing) {
      actions = [
        h('button', { className: 'save', onClick: this.handleSave }, 'Save'),
        h('button', { className: 'cancel', onClick: this.handleCancel }, 'Cancel'),
      ];
    } else if (canEdit) {
      actions = [
        h('button', { className: 'edit', onClick: this.handleEdit }, 'Edit'),
      ];
    }

    let hoodInfo = null;
    if (editing) {
      hoodInfo = h('div', [
        h('label', { htmlFor: 'hood-name' }, 'Hood'),
        h('input', {
          className: 'name-input',
          type: 'input',
          value: name,
          onChange: this.handleInput,
        }),
      ]);
    } else {
      hoodInfo = h('div', name);
    }

    const VoteState = userVoted ? Voted : VoteUp;
    const UserVoting = canUserVote ? h(VoteState, {
      className: 'fa fa-angle-up',
      onClick: () => {
        if (userVoted) {
          handleUnvote(hoodId, currentUserId);
          return;
        }

        handleVote(hoodId, currentUserId);
      },
    }) : null;

    return h(OverlayContainer, [
      h(OverlayHeader, 'Hood View'),
      h(HoodContainer, [
        h(Votes, [
          UserVoting,
          h('div', votes),
        ]),
        h('div', [
          hoodInfo,
          h('div', [
            h('span', 'User: '),
            h('span', user.isTmp ? 'Anonymous' : user.email),
          ]),
        ]),
        actions ? h(Actions, actions) : null,
      ]),
    ]);
  }
}

export default connect(
  (state: State) => {
    const hood = getHoodSelected(state);
    if (!hood) {
      return <DefaultProps>{ hood: null };
    }
    const userId = hood.userId;
    const user = getUserById(state, { id: hood.userId });
    const currentUserId = getCurrentUserId(state);
    const didUserCreateHood = user && user.id === currentUserId;
    const hoodId = hood.id;
    const userIsAuthenticated = isUserAuthenticated(state);
    const canUserVote = userIsAuthenticated;

    return {
      hood,
      hoodId,
      user: user || undefined,
      canEdit: didUserCreateHood,
      votes: getVoteCountByHood(state, { hoodId }),
      userVoted: didUserVoteOnHood(state, { hoodId, userId }),
      canUserVote,
      currentUserId,
    };
  },
  (dispatch: Function) => ({
    updateHoodName: (payload: SetHoodNamePayload) => dispatch(setHoodName(payload)),
    handleDeselectHood: () => dispatch(deselectHood()),
    hideHoodOverlay: () => dispatch(hideMenu('overlay')),
    edit: (opts: EditHoodPayload) => dispatch(editHood(opts)),
    save: (hoodId: HoodId) => dispatch(saveHood(hoodId)),
    handleVote: (hoodId: HoodId, userId: UserId) => dispatch(vote({ hoodId, userId })),
    handleUnvote: (hoodId: HoodId, userId: UserId) => dispatch(unvote({ hoodId, userId })),
  }),
)(HoodView as any);
