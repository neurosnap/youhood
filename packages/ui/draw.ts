import * as h from 'react-hyperscript';
import { connect } from 'react-redux';
import styled from 'styled-components';

import { actions, selectors } from '@youhood/hood';
const { drawHood, cancelDrawHood } = actions;
const { getIsEditing } = selectors;
import { selectors as authSelectors } from '@youhood/auth';
const { isUserAuthenticated } = authSelectors;

import { HoodBarButton } from './ui';

const Button = HoodBarButton.extend`
  width: 91px;
  display: flex;
  justify-content: center;
`;

const Overlay = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  width: 100vw;
  height: 56px;
  background-color: rgba(0, 0, 0, 0);
  cursor: pointer;
`;

interface ICancel {
  cancelDraw: CancelFn;
}
export const Cancel: React.SFC<ICancel> = ({ cancelDraw }) =>
  h('div', [
    h(Button, { onClick: cancelDraw }, [h('div', 'Cancel')]),
    h(Overlay, { onClick: cancelDraw }),
  ]);

type CancelFn = (event: React.MouseEvent<HTMLElement>) => void;
interface IState {
  isUserLoggedIn: boolean;
  isEditing: boolean;
}
interface IDispatch {
  handleCancelDrawHood: CancelFn;
  handleDrawHood: (event: React.MouseEvent<HTMLButtonElement>) => void;
}
interface Props extends IState, IDispatch {}

export const DrawHood: React.SFC<Props> = ({
  handleDrawHood,
  handleCancelDrawHood,
  isEditing,
  isUserLoggedIn,
}) => {
  if (!isUserLoggedIn) {
    return null;
  }

  return isEditing
    ? h(Cancel, { cancelDraw: handleCancelDrawHood })
    : h(Button, { onClick: handleDrawHood }, 'Create Hood');
};

export default connect<IState, IDispatch>(
  (state) => ({
    isEditing: getIsEditing(state),
    isUserLoggedIn: isUserAuthenticated(state),
  }),
  (dispatch) => ({
    handleDrawHood: () => dispatch(drawHood()),
    handleCancelDrawHood: () => dispatch(cancelDrawHood()),
  }),
)(DrawHood);
