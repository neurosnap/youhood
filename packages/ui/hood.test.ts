import { mount } from 'enzyme';
import * as h from 'react-hyperscript';

import { HoodView } from './hood';
import { Voted, VoteUp } from './ui';

const tmpUser = {
  email: '123',
  id: '123',
  isTmp: true,
};
const regUser = {
  email: 'eric@cool.com',
  id: '123',
  isTmp: false,
};

describe('HoodView', () => {
  it('should render null', () => {
    const tree = mount(h(HoodView));
    expect(tree).toMatchSnapshot();
  });

  describe('when the user is temporary', () => {
    it('should render `Anonymous` user name', () => {
      const tree = mount(h(HoodView, {
        show: true,
        hoodId: '321',
        user: tmpUser,
      }));

      expect(tree).toMatchSnapshot();
    });
  });

  describe('when the user has registered', () => {
    it('should render the email as user', () => {
      const tree = mount(h(HoodView, {
        show: true,
        hoodId: '321',
        user: regUser,
      }));

      expect(tree).toMatchSnapshot();
    });
  });

  describe('when the user can vote on hood', () => {
    describe('when the user has not voted yet', () => {
      const handleVote = jest.fn();
      const tree = mount(h(HoodView, {
        show: true,
        hoodId: '321',
        user: regUser,
        userVoted: false,
        canUserVote: true,
        handleVote,
        currentUserId: regUser.id,
      }));
      const vote = tree.find(VoteUp);

      it('should display the upvote arrow', () => {
        expect(vote.length).toEqual(1);
      });

      describe('when the user clicks the upvote arrow', () => {
        it('should call handleVote', () => {
          vote.simulate('click');
          expect(handleVote).toHaveBeenCalledWith('321', regUser.id);
        });
      });
    });

    describe('when the user has voted', () => {
      const handleUnvote = jest.fn();
      const tree = mount(h(HoodView, {
        show: true,
        hoodId: '321',
        user: regUser,
        userVoted: true,
        canUserVote: true,
        handleUnvote,
        currentUserId: regUser.id,
      }));
      const vote = tree.find(Voted);

      it('should display the voted arrow', () => {
        expect(vote.length).toEqual(1);
      });

      describe('when the user clicks the upvote arrow', () => {
        it('should call handleUnvote', () => {
          vote.simulate('click');
          expect(handleUnvote).toHaveBeenCalledWith('321', regUser.id);
        });
      });
    });
  });

  describe('when the user can edit the hood', () => {
    describe('when the user is temporary', () => {
      it('should display edit button', () => {
        const tree = mount(h(HoodView, {
          show: true,
          hoodId: '321',
          user: tmpUser,
          canEdit: true,
        }));

        expect(tree.find('.edit').length).toEqual(1);
      });
    });

    describe('when the user has registered', () => {
      const tree = mount(h(HoodView, {
        show: true,
        hoodId: '321',
        user: regUser,
        canEdit: true,
      }));

      expect(tree.find('.edit').length).toEqual(1);
    });
  });

  describe('when the user is editing hood', () => {
    const edit = jest.fn();
    const tree = mount(h(HoodView, {
      show: true,
      hoodId: '333',
      user: regUser,
      canEdit: true,
      hood: {
        id: '333',
        name: 'kerrytown',
      },
      edit,
    }));

    tree.find('.edit').simulate('click');

    describe('when clicking the edit button', () => {
      it('should set `editing` to true', () => {
        expect(tree.state('editing')).toEqual(true);
      });

      it('should call `edit`', ()  => {
        expect(edit).toHaveBeenCalledWith({ edit: true, hoodId: '333' });
      });
    });

    it('should render save button', () => {
      expect(tree.find('.save').length).toEqual(1);
    });

    it('should render the cancel button', () => {
      expect(tree.find('.cancel').length).toEqual(1);
    });

    it('should render input', () => {
      expect(tree.find('.name-input').length).toEqual(1);
    });

    it('should have the correct hood name', () => {
      expect(tree.find('.name-input').props().value).toEqual('kerrytown');
    });

    describe('when typing in name input', () => {
      it('should set state name to input value', () => {
        const input = tree.find('.name-input');
        input.simulate('change', { target: { value: 'downtown' } });
        expect(tree.state('name')).toEqual('downtown');
      });
    });
  });

  describe('when the user clicks the save button', () => {
    const save = jest.fn();
    const edit = jest.fn();
    const updateHoodName = jest.fn();

    const tree = mount(h(HoodView, {
      show: true,
      hoodId: '333',
      user: regUser,
      canEdit: true,
      hood: {
        id: '333',
        name: 'kerrytown',
      },
      save,
      edit,
      updateHoodName,
    }));

    // simulate editing a hood's name and saving it
    tree.find('.edit').simulate('click');
    const input = tree.find('.name-input');
    input.simulate('change', { target: { value: 'downtown' } });
    tree.find('.save').simulate('click');

    it('should call updateHoodName', () => {
      expect(updateHoodName).toHaveBeenCalledWith({ hoodId: '333', name: 'downtown' });
    });

    it('should set editing state to false', () => {
      expect(tree.state('editing')).toEqual(false);
    });

    it('should call `edit`', () => {
      expect(edit).toHaveBeenCalledWith({ hoodId: '333', edit: false });
    });

    it('should call `save`', () => {
      expect(save).toHaveBeenCalledWith('333');
    });
  });

  describe('when the user clicks the cancel button', () => {
    const edit = jest.fn();

    const tree = mount(h(HoodView, {
      show: true,
      hoodId: '333',
      user: regUser,
      canEdit: true,
      hood: {
        id: '333',
        name: 'kerrytown',
      },
      edit,
    }));

    // simulate editing a hood's name and cancelling it
    tree.find('.edit').simulate('click');
    const input = tree.find('.name-input');
    input.simulate('change', { target: { value: 'downtown' } });
    tree.find('.cancel').simulate('click');

    it('should revert name in state', () => {
      expect(tree.state('name')).toEqual('kerrytown');
    });

    it('should set editing in state to false', () => {
      expect(tree.state('editing')).toEqual(false);
    });

    it('should call `edit`', () => {
      expect(edit).toHaveBeenCalledWith({ hoodId: '333', edit: false });
    });
  });

  describe('when a new hood is selected', () => {
    it('should set name in state to new hood name', () => {
      const tree = mount(h(HoodView, {
        show: true,
        hoodId: '333',
        user: regUser,
        canEdit: true,
        hood: {
          id: '333',
          name: 'kerrytown',
        },
      }));

      tree.setProps({
        hoodId: '222',
        hood: {
          id: '222',
          name: 'wowza',
        },
      });

      expect(tree.state('name')).toEqual('wowza');
    });
  });
});
