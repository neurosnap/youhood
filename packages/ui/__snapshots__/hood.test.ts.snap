// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HoodView should render null 1`] = `
<HoodView
  canEdit={false}
  canUserVote={false}
  edit={[Function]}
  hood={null}
  save={[Function]}
  show={false}
  user={
    Object {
      "createdAt": "",
      "email": "Unknown",
      "id": "",
      "isTmp": true,
    }
  }
  userVoted={false}
  votes={0}
/>
`;

exports[`HoodView when the user has registered should render the email as user 1`] = `
.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: #fff;
  border-radius: 5px;
  padding: 15px;
}

.c3 {
  margin-right: 15px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  font-size: 16px;
  color: #fff;
}

.c0 {
  width: 90%;
  margin-top: 15px;
}

<HoodView
  canEdit={false}
  canUserVote={false}
  edit={[Function]}
  hood={null}
  hoodId="321"
  save={[Function]}
  show={true}
  user={
    Object {
      "email": "eric@cool.com",
      "id": "123",
      "isTmp": false,
    }
  }
  userVoted={false}
  votes={0}
>
  <styled.div>
    <div
      className="c0"
    >
      <styled.div>
        <div
          className="c1"
        >
          Hood View
        </div>
      </styled.div>
      <styled.div>
        <div
          className="c2"
        >
          <styled.div>
            <div
              className="c3"
            >
              <div>
                0
              </div>
            </div>
          </styled.div>
          <div>
            <div />
            <div>
              <span>
                User: 
              </span>
              <span>
                eric@cool.com
              </span>
            </div>
          </div>
        </div>
      </styled.div>
    </div>
  </styled.div>
</HoodView>
`;

exports[`HoodView when the user is temporary should render \`Anonymous\` user name 1`] = `
.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: #fff;
  border-radius: 5px;
  padding: 15px;
}

.c3 {
  margin-right: 15px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  font-size: 16px;
  color: #fff;
}

.c0 {
  width: 90%;
  margin-top: 15px;
}

<HoodView
  canEdit={false}
  canUserVote={false}
  edit={[Function]}
  hood={null}
  hoodId="321"
  save={[Function]}
  show={true}
  user={
    Object {
      "email": "123",
      "id": "123",
      "isTmp": true,
    }
  }
  userVoted={false}
  votes={0}
>
  <styled.div>
    <div
      className="c0"
    >
      <styled.div>
        <div
          className="c1"
        >
          Hood View
        </div>
      </styled.div>
      <styled.div>
        <div
          className="c2"
        >
          <styled.div>
            <div
              className="c3"
            >
              <div>
                0
              </div>
            </div>
          </styled.div>
          <div>
            <div />
            <div>
              <span>
                User: 
              </span>
              <span>
                Anonymous
              </span>
            </div>
          </div>
        </div>
      </styled.div>
    </div>
  </styled.div>
</HoodView>
`;
